README.md
# ADStream module
ADStream makes it easy to configure an areaDetector camera IOC for one or more image
streams, each of which has it's own dedicated stdArray, ROI, CC, Overlay, and Process
plugin, along w/ some extra PV's to characterize the stream type, input source, height,
width, and desired update rate.

You can also do image averaging, and view and adjust 4 crosses and 4 ROI overlay boxes.
The overlays have calc records to handle binning differences between streams.
A python script is used to reconfigure each stream's plugin chain as needed to produce the desired result.

ADStream currently assumes one camera per IOC, w/ camera PORT named CAM, and commonPlugins loaded
to provide other stream input sources.

Stream types include:
* DataStream: Full rate, full res
* ViewingStream: Defaults to 10hz, (configurable), 640x560 target dimensions (configurable), subject to bandwidth limits
* ThumbnailStream: Always B/W,  8 bit pixels, 174x130 target dimensions, default 1hz rate

Stream PV's:
The STREAM prefix for stream PV's is $(CAM):$(IMAGE)
PV example: CAMR:FEE1:1692:IMAGE1:StreamType
* $(STREAM):StreamType	 - Enum: Data, Viewer, or Thumbnail
* $(STREAM):StreamInput	 - Enum: CAM,ROI1,ROI2,ROI3,ROI4,Proc1,Trans1,CC1,CC2
* $(STREAM):StreamHeight - Longout: Desired image height in pixels, 0 to default based on StreamType and StreamInput
* $(STREAM):StreamWidth  - Longout: Desired image width  in pixels, 0 to default based on StreamType and StreamInput
* $(STREAM):StreamRate   - Float: Desired image rate in hz, 0 to default based on StreamType and StreamInput

To enable/disable cross and ROI box overlays:
* $(STREAM):Cross1Use	- Enum: No or Yes
* $(STREAM):Cross2Use	- Enum: No or Yes
* $(STREAM):Cross3Use	- Enum: No or Yes
* $(STREAM):Cross4Use	- Enum: No or Yes
* $(STREAM):Box1Use	    - Enum: No or Yes
* $(STREAM):Box2Use	    - Enum: No or Yes
* $(STREAM):Box3Use	    - Enum: No or Yes
* $(STREAM):Box4Use	    - Enum: No or Yes

Some limited Image processing can be enabled for each stream.
* $(STREAM):Proc:EnableBackground - Enum: Disable or Enable background subtraction
* $(STREAM):Proc:SaveBackground   - Enum: No or Yes, toggle on/off to save a new background image
* $(STREAM):Proc:EnableFilter     - Enum: Disable or Enable image averaging
* $(STREAM):Proc:NumFilter        - Number of images to average

To reconfigure the stream plugins for current StreamInput and other stream PV settings, run the python script:
ADStreamScreens/configStream.py  CameraPVPrefix  ImageName
Ex. ADStreamScreens/configStream.py CAMR:FEE1:1692  IMAGE1

The configStream.py script will automatically insert a color correction plugin if needed,
followed by an ROI plugin if needed to do any binning.
If Image processing is enabled, the Proc plugin comes next.
Then if any crosses or boxes are enabled, the Overlay plugin.
Finally, a stdArray plugin is used to provide the output image.
If no additional plugins are needed, stdArray could get it's input directly from the camera.

Coordinates for the 4 ROI boxes assume you're using the 4 ROI plugins from common plugins, each using CAM as their NDArrayPort.
* $(CAM):ROI1:MinX
* $(CAM):ROI1:MinY
* $(CAM):ROI1:SizeX
* $(CAM):ROI1:SizeY
* $(CAM):ROI2:MinX
...
* $(CAM):ROI4:SizeY

To load a set of 4 cross PV's for your camera, use dbLoadRecords( db/cross.db, CAM=$(CAM) )
* $(CAM):Cross1MinX
* $(CAM):Cross1MinY
* $(CAM):Cross1SizeX
* $(CAM):Cross1SizeY
* $(CAM):Cross2MinX
...
* $(CAM):Cross4SizeY

HW vs SW ROI and Binning:
ADCore supports many ways to handle ROI and binning.
Here is how ADStream does ROI and binning for each stream.
1.  HW ROI and binning
	These PV's are not supported by all drivers, but when available can increase
	frame rates and reduce resource consumption for all plugins.
	$(CAM):BinX		- Camera and/or driver X binning factor
	$(CAM):BinY		- Camera and/or driver Y binning factor
	$(CAM):MinX		- Camera and/or driver X ROI start
	$(CAM):MinY		- Camera and/or driver Y ROI start
	$(CAM):SizeX	- Camera and/or driver X ROI size
	$(CAM):SizeY	- Camera and/or driver Y ROI size
	Each of these has an _RBV equivalent which the driver sets to the
	actual HW ROI and/or binning.
	i.e.  You may Set $(CAM):BinX to 5, but if the camera/driver doesn't
	support that, $(CAM):BinX_RBV may be 1

2.  Software ROI
	ADStream uses the 4 ROI plugins from common plugins to select 4 ROI regions using
	camera pixel coordinates after any HW ROI or binning have been applied.
	$(CAM):ROI1:MinX	- Software ROI1 X start
	$(CAM):ROI1:MinY	- Software ROI1 Y start
	$(CAM):ROI1:SizeX	- Software ROI1 X size
	$(CAM):ROI1:SizeY	- Software ROI1 Y size
	$(CAM):ROI2:MinX	- Software ROI2 X start
	...
	$(CAM):ROI4:SizeY	- Software ROI4 Y size
	Again each has an _RBV equivalent which the plugin sets to the actual readback value.
	ADStream doesn't use the binning available via these 4 ROI plugins from commonPlugins.

3.  Software binning
	ADStream allows each stream to do it's own binning
	$(STREAM):ROI:BinX	- Set by configStream.py based on StreamInput and StreamType
	$(STREAM):ROI:BinY	- Set by configStream.py based on StreamInput and StreamType
	$(STREAM):ROI:MinX	- Set by configStream.py based on StreamInput and stream binning
	$(STREAM):ROI:MinY	- Set by configStream.py based on StreamInput and stream binning
	$(STREAM):ROI:SizeX	- Set by configStream.py based on StreamInput and stream binning
	$(STREAM):ROI:SizeY	- Set by configStream.py based on StreamInput and stream binning
	Again each has an _RBV equivalent which the plugin sets to the actual readback value.
	These values can be set manually as well, but may get stepped on the next time
	configStream.py runs.

As is normal for ADCore, the image width and height for every ArrayData waveform produced
by StdArray plugins are directly available via these PV's:
	$(STREAM):ArraySize0_RBV
	$(STREAM):ArraySize1_RBV
For color RGB images, ArraySize0_RBV will be 3 and the image width and height are in
ArraySize1_RBV and ArraySize2_RBV respectively.


Building ADStream:

ADStream depends on ADCore, so you'll need to define the path to your ADCore release in
configure/RELEASE or configure/RELEASE.local.
ADStream has been tested w/ ADCore versions R2.0, R2.1, and R2.6
but should work w/ newer versions as well.

## Files built in db directory
ADStream populates the db directory w/ a combination of db files and cmd files.
The cmd files are iocsh scripts to facilitate loading the appropriate db file or files and configuring any
plugins which it loads.  Epics env variables are used to configure PV names and characteristics
for the plugins, and the required ENV variables are documented in each cmd file.

The db files are produced via substitution files, so each ADCore plugin type has a db file and a cmd file.

There are also cmd scripts and db files for each type of stream.

To use in an IOC, define the ADSTREAM macro in one of your RELEASE files and add the following lines
in your App/Db/Makefile:
```
DB_INSTALLS += $(wildcard $(ADSTREAM)/db/*.cmd)
DB_INSTALLS += $(wildcard $(ADSTREAM)/db/*.db)
```

In your st.cmd file, you'll need to define macros for the camera base prefix and port.
For example:
```
epicsEnvSet( "CAM",      "TST:GIGE:BASLER2" )
epicsEnvSet( "CAM_PORT", "CAM" )
```

Most of the remaining required env variables characterize the camera characteristics:
* IMAGE_XSIZE
* IMAGE_YSIZE
* IMAGE_NELM
* IMAGE_BIT_DEPTH
* IMAGE_TYPE
* IMAGE_FTVL
* STREAM_NELM

STREAM_NELM is often set the same as IMAGE_NELM but can be reduced if desired
to ensure the data transfered via EPICS CA isn't too large even if variable sized
arrays aren't supported.

We typically define these by sourcing an env file specific to the camera model.
For example, our MantaG146C.env file, which we've added to our ADProsilica module,
looks like this:
```
#
# ----- Manta G146C ENV Vars -----
#
epicsEnvSet( "IMAGE_FTVL",      "UCHAR"     )
epicsEnvSet( "IMAGE_TYPE",      "Int8"      )
epicsEnvSet( "IMAGE_BIT_DEPTH", "12"        )
epicsEnvSet( "IMAGE_XSIZE",     "1388"      )
epicsEnvSet( "IMAGE_YSIZE",     "1038"      )
epicsEnvSet( "IMAGE_NELM",      "4322232"   )   # X*Y (B/W) or X*Y*3 (Color)
```

In st.cmd, you can then install as many image streams as desired by defining the common env variables
and then adding lines like this:
```
# Add DataStream named DATA1
epicsEnvSet( "IMAGE_NAME", "DATA1" )
. db/DataStream.cmd

# Add ViewerStreams named IMAGE1 and IMAGE2
epicsEnvSet( "VIEWER_NELM", $(IMAGE_NELM) )
epicsEnvSet( "IMAGE_NAME", "IMAGE1" )
. db/ViewerStream.cmd
epicsEnvSet( "IMAGE_NAME", "IMAGE2" )
. db/ViewerStream.cmd

# Add a ThumbnailStream named THUMBNAIL
epicsEnvSet( "THUMBNAIL_NELM", $(IMAGE_NELM) )
epicsEnvSet( "IMAGE_NAME", "THUMBNAIL" )
. db/ThumbnailStream.cmd
```

Individual plugins can be loaded by defining macros N and PLUGIN\_SRC, along w/ any plugin specific macros,
and sourcing the appropriate cmd script.
For example, you can load an additional Stats plugin like this:
```
epicsEnvSet( "N", 2 )
epicsEnvSet( "PLUGIN_SRC", "$(CAM_PORT)" )
. db/pluginStats.cmd
```

## EDM usage
From edm, the way it works is that any buttons that change stream characteristics are stacked
on top of shell widgets that run the configStream.py script, which is found in the same
directory as the edm screens and located at run-time via a soft link ADStreamScreens/configStream.py.
The script works it's way through the stream's plugins, setting NDArrayPort, plugin enables,
minCallbackTime, binning, etc as needed to  produce the desired result with only the needed
plugins in the chain.

There are also buttons to create a new custom edm screen and switch to it if the stdArray
output dimensions change.

The 4 ROI regions and 4 cross locations and sizes are shared between all image streams and
configured using unbinned camera pixels.    Overlays to show these are scaled down by the
appropriate bin factor and selected region for each image stream. Each of the cross and ROI box
overlays can be turned on and off independently for each stream. Image averaging and color
conversion can also be configured independently for each stream.


RELEASE_NOTES
#==============================================================
#
#  Name: RELEASE_NOTES
#
#==============================================================
R3.1.2	2023-02-21 Bruce Hill
	Updated to ADCore/R3.10-1.1.0

R3.1.1	2022-12-02 Bruce Hill
	Replace TPS_PV w/ TPE_PV.

R3.1.0	2022-11-28 Bruce Hill
	Support TPR or EVR in ADStream screens

R3.0.2	2022-02-09 Bruce Hill
	Update to ADCore/R3.9-1.1.0

R3.0.1	2022-01-28 Bruce Hill
	Update to ADCore/R3.9-1.0.3

R3.0.0	2021-10-29 Bruce Hill
	Fix NDTimeSeriesConfigure() parameters
	Default NDArrayAddress VAL should be 0
	Built w/ ADCore/R3.9-1.0.2

R2.9.3 2022-01-20 Bruce Hill
	Update to ADCore/R3.9-1.0.3

R2.9.2 2021-05-26 Bruce Hill
	Update to ADCore/R3.9-1.0.2

R2.9.1 2021-04-27 Bruce Hill
	Update to ADCore/R3.9-1.0.1

R2.9.0 2021-04-20 Bruce Hill
	Update to ADCore/R3.9-1.0.0

R2.8.0 2021-02-04 Bruce Hill
	Set HIGH alarm for non-zero DataStream delay.

R2.7.5  2020-11-11 Bruce Hill
	Add test for valid screen path in EDM_TOP env var
	Tweak ImagePlugin default colors
	Adding support for LCLS config in edmViewer.sh

R2.7.4  2020-10-02 Bruce Hill
	Update to ADCore/R3.1-2.3.5

R2.7.3  2020-05-20 Bruce Hill
	Create a release for ADCore/R3.1-2.3.1

R2.7.2		2020-03-03 Bruce Hill
	Update to ADCore/R3.1-2.3.4

R2.7.1      2019-06-14 Bruce Hill
	Removed unwanted macros from RELEASE.local

R2.7.0      2019-06-14 Bruce Hill
	Cut st.cmd output by pre-building commonPlugins.db

R2.6.3      2019-06-12 Bruce Hill
	Update to ADCore/R3.1-2.3.3
	Change comments to begin w/ #- to inhibit iocsh echoing

R2.6.2      2018-11-30 Bruce Hill
	Update to ADCore/R3.1-2.3.0

R2.6.1      2018-8-28 Bruce Hill
	Use shorter timestamp format for StreamTimeStampStr

R2.6.0      2018-8-28 Bruce Hill
	Modified plugin config calls to allow specifing MAX_THREADS
	Add a StreamTimeStampStr record w/ a formatted timeStamp for gui display
	Update to ADCore/R3.1-2.2.0

R2.5.4      2018-3-19 Bruce Hill
	Updated to ADCore/R3.1-2.1.1

R2.5.3      2018-3-19 Bruce Hill
	Updated to ADCore/R3.1-2.1.0 for qsrv fix

R2.5.2      2018-1-29 Bruce Hill
	Updated to ADCore/R3.1-2.0.0

R2.5.1      2018-1-29 Bruce Hill
	Updated to ADCore/R3.1-1.0.0

R2.5.0      2018-1-22 Bruce Hill
	Fixed missing CH macro in emb-thumbnail.edl

R2.4.0      2017-11-20 Bruce Hill
	Added a V4 NDPva plugin to each stream type.
	Lengthened default timeout from 0.1 to 0.5 sec to reduce failed connections.
	Fixed a number of issues related to color cameras.
	Made ROI scaling only be used as needed to prevent exceeding max bitsPerPixel
	Changed some defaults so initial bootup has reasonable stream plugin processing.
	Cleanup box and cross limits and overlay substitutions.
	Update to ADCore/R2.6-2.4.0

R2.3.0      2017-08-22 Bruce Hill
	Added edmLaunch.sh that takes a CAM prefix, fetches $CAM:LAUNCH_EDM and launchs the edm top screen.

R2.2.0      2017-08-09 Bruce Hill
	Added CH macro to launcher and ViewerTemplate's

R2.1.0      2017-08-07 Bruce Hill
	Auto set of BayerConvert, Proc:EnableCallbacks, and Over:EnableCallbacks
	Made related displays popup relative to button position
	Added optional macros w/ defaults for pluginStats:
		HIST_SIZE, NCHANS, RATE_SMOOTH, and TSE
	Fixed IMAGE_NAME in thumbnailStream to allow names other than THUMBNAIL
	Removing obsolete pcdsPlugins and Mono*Viewer.cmd files
	Added more documentation to README.md including
		details on stream pv's, HW vs SW ROI and binning,
		and python configStream.py script.
	Fixed pv names for cross color.
	Added Red, Green, and Blue pv's and made Color map to Green for mono overlays.
	Opened up constraints on viewerStreams

R2.0.0	2017-06-16 Bruce Hill
	Fix typo in variable name
	Added a configurable STREAM_NELM so we can force streams
		to have a smaller NELM than the image.
	Add check and auto-enable for BayerConvert turned off

R1.9.2      2017-05-18 Hugo Slepicka
    Updated to latest ADCore/R2.6-2.2.0

R1.9.1      2017-05-17 Hugo Slepicka
    Updated to latest ADCore/R2.6-2.1.3

R1.9.0		2017-05-03 Bruce Hill
	Updated to latest ADCore/R2.6-2.1.0
	Fixed numerous substitutions issues related to NDPluginBase.template
		now being included in each plugin template file.

R1.8.1		2017-04-20 Bruce Hill
	Updated for ADCore/R2.6-0.3.0

R1.8.0		2017-04-12 Bruce Hill
	Updated for compatibility w/ ADCore/R2.6-0.2.0
	(macro name Enabled changed to ENABLED) 

R1.7.1		2017-01-31 Bruce Hill
	Updated ADCore to R2.2-1.1.1

R1.7.0		2017-01-31 Bruce Hill
	Fixed visibility of BayerMode vs BitsPerPixel in emb-stream-sts.edl
	Problem was due to nested visibility groups based on the same PV w/ different visibility settings.
		edm does not support this so reworked visibility groups on ColorMode to avoid nesting.
	Added comment re python script
	Broke out error checking of RELEASE_SITE and EPICS_BASE to TOP/../../RELEASE_SITE.check
		and TOP/../../EPICS_BASE.check

R1.6.2		2016-10-14 Bruce Hill
	Updated ADCore to R2.2-1.0.0
	Updated build files for git module work-flow

R1.6.1		2016-10-14 Bruce Hill
	Updated build files for git module work-flow

ADStreamScreens/configStream.py
#!/bin/env python
##############################################################################
## This file is part of 'ADStream'.
## It is subject to the license terms in the LICENSE.txt file found in the 
## top-level directory of this distribution and at: 
##    https://confluence.slac.stanford.edu/display/ppareg/LICENSE.html. 
## No part of 'ADStream', including this file, 
## may be copied, modified, propagated, or distributed except according to 
## the terms contained in the LICENSE.txt file.
##############################################################################
# pyca script for python 2.7.2

# TODO: ROI being selected for color images when it shouldn't be
# Set stream height and width to defeat ROI, but Scaling was left enabled and ROI plugin still in path
# Also issues w/ OVER added to path for IMAGE2

from psp.Pv import Pv
import pyca
import sys
import psp

from psp.options import Options

showCAErrors		= False
TY_STREAM_DATA		= 0
TY_STREAM_VIEWER	= 1
TY_STREAM_THUMBNAIL	= 2

def printPvNameValue( pvName ):
    try:
        pv = Pv( pvName )
        pv.connect(0.5)
        pv.get()
        pv.get_data( False, 0.5, None )
        if isinstance( pv.value, str ):
            print "%s \"%-.30s\"" % ( pv.name, pv.value )
        else:
            print "%s %-.30s" % ( pv.name, pv.value )
    except pyca.pyexc, msg:
        if showCAErrors:
            print >> sys.stderr, "failed: pyca exception: ", msg
        pass
    except pyca.caexc, msg:
        print >> sys.stderr, "failed: channel access exception: ", msg
    except Exception, msg:
        print >> sys.stderr, "failed:", msg.__class__.__name__, "exception: ", msg

def tupleToString( tupleValue ):
    if isinstance( tupleValue, str ):
        return tupleValue
    # Remove any invalid characters
    for item in tupleValue:
        if item > 255:
            tupleValue.remove(item)

    # Isolate the null terminated string, if any
    if tupleValue.count(0) == 0:
        end = len(tupleValue)
    else:
        end = tupleValue.index(0)

    # Convert remaining bytes to a string
    stringRet = ''.join( map( chr, tupleValue[0:end] ) )
    return stringRet

def stringToTuple( strValue ):
    if not isinstance( strValue, str ):
        raise TypeError, "stringToTuple value is not a string"

    # Convert the string to a bytearray
    newByteArray = bytearray( strValue, 'ascii', 'ignore' )
    return tuple(newByteArray)

def caGetValue( pvName, default=None, verbose=True, timeout=0.5 ):
    try:
        # See if this PV exists
        pv	= Pv( pvName )
        pv.connect( timeout )
        pv.get( timeout=timeout )
        pv.get_data( False, timeout, None )
        return pv.value
    except Exception, msg:
        if verbose:
            print "Unable to connect to PV: %s" % pvName
        if showCAErrors:
            print >> sys.stderr, "failed: pyca exception: ", msg
        return default


def caPutArray( pvName, value, timeout=1.0):
    try:
        tmo = float(timeout)
        # See if this PV exists
        pv	= Pv( pvName )
        pv.connect( 1.0 )
        pv.get_data( False, timeout, None )

        # Convert value to the appropriate type for this PV
        # Assumption is that PV is a waveform record
        if isinstance( value, str ):
            # Convert string values to a tuple
            value = stringToTuple( value )
        if not isinstance( value, tuple ):
            raise TypeError, "Pv.put Error: value not a tuple for waveform PV"
        if len(value) != len(pv.value):
            raise TypeError, "Pv.put Error: Element count mismatch"

        pv.put_data(value, tmo)

    except Exception, msg:
        if verbose:
            print "Unable to connect to PV:", pvName
        if showCAErrors:
            print >> sys.stderr, "failed: pyca exception: ", msg
        return

def caPutValue( pvName, value, verbose=True ):
    try:
        # See if this PV exists
        pv	= Pv( pvName )
        pv.connect( 1.0 )
        pv.put( value, timeout=1.0 )
    except Exception, msg:
        if verbose:
            print "Unable to connect to PV:", pvName
        if showCAErrors:
            print >> sys.stderr, "failed: pyca exception: ", msg
        return

def reconfigStream( cameraPvName, streamName, verbose=False ):
    streamPvName	= cameraPvName + ":" + streamName
    try:
        streamTypePv = Pv( streamPvName + ":StreamType" )
        streamTypePv.connect(0.5)
        streamTypePv.get( timeout=0.5 )
        streamTypePv.get_data( False, 0.5, None )
        streamType = streamTypePv.value
    except Exception, msg:
        if verbose:
            print "Stream %s not found." % streamName
        return

    # Fetch the stream's type and input port
    streamType		= caGetValue( streamPvName + ":StreamType" )
    streamPort		= caGetValue( streamPvName + ":StreamPort" )
    streamNELM		= caGetValue( streamPvName + ":ArrayData.NELM" )
    upStreamPort	= "CAM"
    if streamPort:
        upStreamPort = streamPort

    # Determine the sourcePvName
    if upStreamPort	== "CAM":
        sourcePvName = cameraPvName
        pluginType	= None
    else:
        # This stream's input is another plugin
        sourcePvName	= cameraPvName + ":" + upStreamPort

        # Make sure the plugin they asked for is enabled
        caPutValue( sourcePvName + ":EnableCallbacks", 1 )
        pluginType = caGetValue( sourcePvName + ":PluginType_RBV", verbose=False )

    sourceColor		= caGetValue( sourcePvName + ":ColorMode_RBV" )
    if pluginType == "NDPluginProcess":
        if sourceColor == 0:
            sourceWidth		= caGetValue( sourcePvName + ":ArraySize0_RBV" )
            sourceHeight	= caGetValue( sourcePvName + ":ArraySize1_RBV" )
        else:
            sourceWidth		= caGetValue( sourcePvName + ":ArraySize1_RBV" )
            sourceHeight	= caGetValue( sourcePvName + ":ArraySize2_RBV" )
    else:
        sourceWidth		= caGetValue( sourcePvName + ":ArraySizeX_RBV" )
        sourceHeight	= caGetValue( sourcePvName + ":ArraySizeY_RBV" )
    sourceBits		= caGetValue( sourcePvName + ":BitsPerPixel_RBV" )

    streamWidth		= caGetValue( streamPvName + ":StreamWidth" )
    streamHeight	= caGetValue( streamPvName + ":StreamHeight" )
    streamRate		= caGetValue( streamPvName + ":StreamRate" )
    avgEnabled		= caGetValue( streamPvName + ":Proc:EnableFilter", verbose=False )
    ccEnabled		= False # TODO: caGetValue( streamPvName + ":UseCC" )
    cross1Enabled	= caGetValue( streamPvName + ":Cross1:Use", default=False )
    cross2Enabled	= caGetValue( streamPvName + ":Cross2:Use", default=False )
    cross3Enabled	= caGetValue( streamPvName + ":Cross3:Use", default=False )
    cross4Enabled	= caGetValue( streamPvName + ":Cross4:Use", default=False )
    box1Enabled		= caGetValue( streamPvName + ":Box1:Use",   default=False )
    box2Enabled		= caGetValue( streamPvName + ":Box2:Use",   default=False )
    box3Enabled		= caGetValue( streamPvName + ":Box3:Use",   default=False )
    box4Enabled		= caGetValue( streamPvName + ":Box4:Use",   default=False )
    overEnabled     = ( cross1Enabled | cross2Enabled | cross3Enabled | cross4Enabled |
                        box1Enabled   | box2Enabled   | box3Enabled   | box4Enabled   )

    if streamType == TY_STREAM_DATA:
        defCallbackTime = 0.0
        minCallbackTime = 0.0
        defStreamWidth	= sourceWidth
        defStreamHeight = sourceHeight
        maxStreamWidth	= sourceWidth
        maxStreamHeight = sourceHeight
        if sourceColor == 0:
            maxBits			= 16
        else:
            maxBits			= 24
        monoOnly		= False
        # TODO: Need to revisit rate vs size in terms of bandwidth
        # May need to make developer provide overrides to get
        # both full resolution and full rate
    elif streamType == TY_STREAM_THUMBNAIL:
        defCallbackTime = 0.9
        minCallbackTime = 0.2
        defStreamWidth	= 174
        defStreamHeight = 130
        maxStreamWidth	= min( 200, sourceWidth )
        maxStreamHeight = min( 200, sourceHeight )
        maxBits			= 8
        monoOnly		= True
    else: # streamType == TY_STREAM_VIEWER
        defCallbackTime = 0.1
        minCallbackTime = 0.03333	# Max 30hz
        defStreamHeight = 600
        defStreamWidth	= 800
        # TODO: Need to revisit rate vs size in terms of bandwidth
        # Viewing an 1400x1400 12bit image at 30hz is 99% of gigabit capacity
        # Max 1600x1600 for VIEWER
        maxStreamWidth	= min( 1600, sourceWidth )
        maxStreamHeight = min( 1600, sourceHeight )
        if sourceColor == 0:
            maxBits			= 16
        else:
            maxBits			= 24
        monoOnly		= False

    # Check against streamNELM
    if  maxStreamWidth * maxStreamHeight > streamNELM:
        maxStreamWidth	= int(streamNELM / 2)
        maxStreamHeight = int(streamNELM / 2)

    # Set callbackTime
    callbackTime = defCallbackTime
    if streamRate > 0:
        callbackTime = 1 / streamRate
    if  callbackTime < minCallbackTime:
        callbackTime = minCallbackTime

    # Set streamHeight
    if not streamHeight:
        streamHeight = defStreamHeight
    if  streamHeight > maxStreamHeight:
        streamHeight = maxStreamHeight

    # Set streamWidth
    if not streamWidth:
        if sourceHeight:
            streamWidth = int(sourceWidth * (float(streamHeight) / sourceHeight))
        else:
            streamWidth = defStreamHeight
    if  streamWidth > maxStreamWidth:
        streamWidth = maxStreamWidth

    if verbose:
        print "%s image is %d x %d, target %d x %d" % ( streamPort,
                sourceWidth, sourceHeight, streamWidth, streamHeight )

    binning = 1
    if sourceWidth > streamWidth or sourceHeight > streamHeight:
        xRatio = float(sourceWidth)  / streamWidth
        yRatio = float(sourceHeight) / streamHeight

        # If ratio is between 1.1 and 2.1, bin by 2
        binX = int( round( xRatio + 0.4) )
        binY = int( round( yRatio + 0.4) )
        binning = max( binX, binY )
        if verbose:
            print "Source/Stream width ratio = %f, height ratio = %f, binning %dx%d" % ( xRatio, yRatio, binning, binning )

    # Check for Bayer mode conversion turned off
    if sourceColor == 1: # Bayer
        caPutValue( cameraPvName + ":BayerConvert", 1 ) # RGB1

    # Start w/ ColorCorrect plugin
    if ccEnabled or ( monoOnly and sourceColor >= 1 ):
        # Use CC
        caPutValue( streamPvName + ":CC:EnableCallbacks", 1 )
        caPutValue( streamPvName + ":CC:NDArrayPort", upStreamPort )
        ccOut = caGetValue( streamPvName + ":CC:ColorModeOut_RBV" )
        if ccOut == 0:
            sourceBits	 = 8
        upStreamPort = streamName + ":CC"
    else:
        caPutValue( streamPvName + ":CC:EnableCallbacks", 0 )

    if not sourceBits:
        sourceBits = 16
    tgtBits = maxBits

    if binning > 1 or tgtBits < sourceBits:
        # Use ROI
        caPutValue( streamPvName + ":ROI:EnableCallbacks", 1 )
        caPutValue( streamPvName + ":ROI:NDArrayPort", upStreamPort )
        caPutValue( streamPvName + ":ROI:EnableX", 1 )
        caPutValue( streamPvName + ":ROI:EnableY", 1 )
        caPutValue( streamPvName + ":ROI:BinX", binning )
        caPutValue( streamPvName + ":ROI:BinY", binning )
        if  tgtBits == 8:
            caPutValue( streamPvName + ":ROI:DataTypeOut", "UInt8" )
        else:
            caPutValue( streamPvName + ":ROI:DataTypeOut", "Automatic" )

        scale = binning * binning
        if sourceBits > tgtBits:
            scale = scale * ( 2 ** (sourceBits - tgtBits) )
        if verbose:
            print "Binning %ux%u, sourceBits = %d, tgtBits = %d, scale = %f" % ( binning, binning, sourceBits, tgtBits, scale )
        if scale >= 2:
            caPutValue( streamPvName + ":ROI:EnableScale", 1 )
            caPutValue( streamPvName + ":ROI:Scale", scale )
        else:
            caPutValue( streamPvName + ":ROI:EnableScale", 0 )
        upStreamPort = streamName + ":ROI"
    else:
        caPutValue( streamPvName + ":ROI:EnableCallbacks", 0 )
        caPutValue( streamPvName + ":ROI:BinX", 1 )
        caPutValue( streamPvName + ":ROI:BinY", 1 )

    avgNum = 1
    if avgEnabled:
        # Use Recursive Averaging Filter in Process plugin
        avgNum		= caGetValue( streamPvName + ":Proc:NumFilter" )
    if avgNum > 1:
        caPutValue( streamPvName + ":Proc:EnableCallbacks", 1 )
        caPutValue( streamPvName + ":Proc:NDArrayPort", upStreamPort )
        upStreamPort = streamName + ":Proc"
    else:
        caPutValue( streamPvName + ":Proc:EnableCallbacks", 0 )

    if overEnabled:
        # Use Overlays
        caPutValue( streamPvName + ":Over:EnableCallbacks", 1 )
        caPutValue( streamPvName + ":Over:NDArrayPort", upStreamPort )
        upStreamPort = streamName + ":Over"
    else:
        caPutValue( streamPvName + ":Over:EnableCallbacks", 0 )

    # Make sure we don't have zero size for ROI and crosses
    if  caGetValue( cameraPvName + ":ROI1:SizeX_RBV" ) == 0:
        caPutValue( cameraPvName + ":ROI1:SizeX", 10 )
    if  caGetValue( cameraPvName + ":ROI1:SizeY_RBV" ) == 0:
        caPutValue( cameraPvName + ":ROI1:SizeY", 10 )
    if  caGetValue( cameraPvName + ":ROI2:SizeX_RBV" ) == 0:
        caPutValue( cameraPvName + ":ROI2:SizeX", 10 )
    if  caGetValue( cameraPvName + ":ROI2:SizeY_RBV" ) == 0:
        caPutValue( cameraPvName + ":ROI2:SizeY", 10 )
    if  caGetValue( cameraPvName + ":ROI3:SizeX_RBV" ) == 0:
        caPutValue( cameraPvName + ":ROI3:SizeX", 10 )
    if  caGetValue( cameraPvName + ":ROI3:SizeY_RBV" ) == 0:
        caPutValue( cameraPvName + ":ROI3:SizeY", 10 )
    if  caGetValue( cameraPvName + ":ROI4:SizeX_RBV" ) == 0:
        caPutValue( cameraPvName + ":ROI4:SizeX", 10 )
    if  caGetValue( cameraPvName + ":ROI4:SizeY_RBV" ) == 0:
        caPutValue( cameraPvName + ":ROI4:SizeY", 10 )
    if  caGetValue( cameraPvName + ":Cross1SizeX" ) == 0:
        caPutValue( cameraPvName + ":Cross1SizeX", 10 )
    if  caGetValue( cameraPvName + ":Cross1SizeY" ) == 0:
        caPutValue( cameraPvName + ":Cross1SizeY", 10 )
    if  caGetValue( cameraPvName + ":Cross2SizeX" ) == 0:
        caPutValue( cameraPvName + ":Cross2SizeX", 10 )
    if  caGetValue( cameraPvName + ":Cross2SizeY" ) == 0:
        caPutValue( cameraPvName + ":Cross2SizeY", 10 )
    if  caGetValue( cameraPvName + ":Cross3SizeX" ) == 0:
        caPutValue( cameraPvName + ":Cross3SizeX", 10 )
    if  caGetValue( cameraPvName + ":Cross3SizeY" ) == 0:
        caPutValue( cameraPvName + ":Cross3SizeY", 10 )
    if  caGetValue( cameraPvName + ":Cross4SizeX" ) == 0:
        caPutValue( cameraPvName + ":Cross4SizeX", 10 )
    if  caGetValue( cameraPvName + ":Cross4SizeY" ) == 0:
        caPutValue( cameraPvName + ":Cross4SizeY", 10 )

    caPutValue( streamPvName + ":EnableCallbacks", 1 )
    caPutValue( streamPvName + ":NDArrayPort",			upStreamPort )
    caPutValue( streamPvName + ":MinCallbackTime",		callbackTime )
    caPutValue( streamPvName + ":CC:MinCallbackTime",	callbackTime )
    caPutValue( streamPvName + ":ROI:MinCallbackTime",	callbackTime )
    caPutValue( streamPvName + ":Over:MinCallbackTime",	callbackTime )

# ----------------------------------------------------------------------
if __name__ == "__main__":

    options = Options( ['cameraPv', 'stream'], [], ['verbose'] )
    try:
        options.parse()
    except Exception, msg:
        options.usage( str(msg) )
        sys.exit()

    cameraPvName = options.cameraPv
    streamName	 = options.stream
    try:
        camSizeXPv = Pv( cameraPvName + ":ArraySizeX_RBV" )
        camSizeXPv.connect(1.0)
        camSizeXPv.get( timeout=1.0 )
        #camSizeX = camSizeXPv.get_data( False, 1.0, None )
    except Exception, msg:
        print "Camera not accessible: ", msg
        sys.exit()

    verbose = False
    if options.verbose is not None:
        verbose = True
    if streamName != 'all':
        reconfigStream( cameraPvName, streamName, verbose=verbose )
    else:
        showCAErrors = False
        reconfigStream( cameraPvName, "DATA1",		verbose=verbose )
        reconfigStream( cameraPvName, "IMAGE1",		verbose=verbose )
        reconfigStream( cameraPvName, "IMAGE2",		verbose=verbose )
        reconfigStream( cameraPvName, "THUMBNAIL",	verbose=verbose )
        reconfigStream( cameraPvName, "DATA2",		verbose=verbose )
        reconfigStream( cameraPvName, "IMAGE3",		verbose=verbose )

    # Done
